### PR TITLE
Implementation of migration area score

### LINKED ISSUE #1734: Discussion about adding a new tally for CMM
Hi all, I create this issue for discussing a few technical details about the implementation pertaining to the Cumulative Migration Method (CMM) for computing diffusion coefficients & transport cross sections (XS). The theory of CMM is still under development and facing a few challenges, but the feature of generating infinite homogenized diffusion coefficients for lattice calculation has been proven rigorous and accurate, and paves the way for fully heterogeneous transport XS. In addition, this feature might also be useful for researchers using OpenMC as requested in issue #1541 .

Early back to 2018 I gave a try to implement this feature of CMM in an older version of OpenMC with mixed codes of fortran & C++, but didn’t keep it up to date with the development of OpenMC. Now thanks to Dr. Jiankai Yu @rockfool who is working on implementing it in the latest openmc source code, we can start to discuss about some details which need more pondering.

Instead of directly adding tally scores for “diffusion coefficients” or “transport XS”, we would rather add a new tally about particles’ movement: the migration area. The migration area is one sixth of the mean square of crow flight length, which is more often called “MSD” (mean square displacement) in fields of particle transport. So in fact the new tally score can also be MSD.

A big improvement in the development of CMM is that now the tally of migration area or MSD can be done in a pure group-wise way, instead of the cumulative-group way when it was first proposed. 

But there are also some limitations:
- Boundary condition
Currently only reflective and vacuum boundary condition are supported for the tally. In reflective boundary conditions, phantom positions of neutrons’ birth places are recorded after each reflection to simulate neutrons’ movement in infinite medium. For vacuum boundary conditions, the tally of MSD will be terminated at the boundary, as if it enters to a material with infinite large absorption XS.
- There can only be one spatial tally zone (the whole geometry)
I made some effort trying to make CMM work in spatially heterogeneous tallies (for cell or mesh filters), but it turned out that there is still more research to be done for pure heterogeneous CMM tallies. So currently the feature we are going to implement only allows for the tally for the whole geometry (as migration area is a quantity defined for the whole core), and no cell/mesh filters are allowed in the tally object for MSD. But energy filter can be used for group-wise tallies.
- Processed results for diffusion coefficients & transport XS
Diffusion coefficients & transport XS are what researchers doing lattice calculations want to get using the CMM feature, and in the current scenario the infinite-homogenized diffusion coefficients can be computed using (D=MSD/6/flux). This works for multi-group tallies and has been tested in many problems. So it might be handy to add a tally score of “diffusion coefficients” as well (in the python API), which can be post-computed using "raw" tally results of MSD and flux.

With the aforementioned limitations, I think firstly we need to talk about the questions below:

1. Proper name for the new tally score: migration area, or MSD (full-spelling might be too long);

2. Warnings for other boundary conditions: can we add warnings when users use MSD tally with periodic/white boundary conditions? If so, should the code report error and terminate?

3. Warnings for spatial tally filters: can we add warnings when user define a tally object containing MSD with cell/mesh filters? If so, should the code report error and terminate?

Looking forward to comments and suggestions on this, thanks!

### PR DESCRIPTION
**Pull request by GuySten**
# Description

This PR implement 'migration-area' score to calculate diffusion coefficients and transport cross sections for multigroup.

Fixes #1734

# Theory
https://www.researchgate.net/publication/324831861_Group-wise_Tally_Scheme_of_Incremental_Migration_Area_for_Cumulative_Migration_Method

# Example

```code
import openmc
import matplotlib.pyplot as plt

model = openmc.Model()

# Material
material = openmc.Material(name="Hydrogen")
material.add_nuclide("H1", 1.0)
material.set_density('g/cm3', 1.0)
material.add_s_alpha_beta('c_H_in_H2O')

# Geometry
radius = 10.0
sphere = openmc.Sphere(r=radius, boundary_type="reflective")
cell = openmc.Cell(region=-sphere, fill=material)
model.geometry = openmc.Geometry([cell])

# Settings
model.settings.particles = 100000
model.settings.batches = 20
model.settings.run_mode = 'fixed source'
model.settings.source = openmc.IndependentSource()
    

# Tally 
groups = openmc.mgxs.EnergyGroups('CASMO-70')
energy_filter = openmc.EnergyFilter(groups.group_edges)
tally = openmc.Tally()
tally.scores = ["migration-area", "flux", "total"]
tally.filters = [energy_filter]
model.tallies = [tally]

model.run()


with openmc.StatePoint('statepoint.20.h5') as sp:
    tally = sp.get_tally(id=1)
    
energies = 0.5*(groups.group_edges[1:]+groups.group_edges[:-1])
flux =  tally.get_values(value='mean', scores=['flux']).squeeze()
total =  tally.get_values(value='mean', scores=['total']).squeeze()
migration_area = tally.get_values(value='mean', scores=['migration-area']).squeeze()

migration_area /= flux
total /= flux

plt.semilogx(energies,1/(3*migration_area*total))
plt.title(r'$\Sigma_{tr}/\Sigma_t$')
plt.show()
```
Figure_1

Reference results:

Untitled

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: docs/source/usersguide/tallies.rst [hunk 1]
```diff
@@ -340,6 +340,15 @@ The following tables show all valid scores:
...
     |                      |probability (IFP) method.                          |
     +----------------------+---------------------------------------------------+
+    |migration-area        |The migration-area score is used to evaluate       |
+    |                      |diffusion coefficients and transport cross sections|
+    |                      |in infinite geometry.                              |
+    |                      |The migration area score can only be used with     |
+    |                      |a ParticleFilter or with an EnergyFilter.          |
+    |                      |It also cannot be used in the same run as a        |
+    |                      |MeshBornFilter for non vacuum boundary conditions. |
+    |                      |For more information, see Liu_.                    |
```
Comment (GiudGiud) [inline docs/source/usersguide/tallies.rst]:
```suggestion
    |                      |MeshBornFilter when non vacuum               |
    |                      |boundary conditions are present.                |
    |                      |For more information, see Liu_.                    |
```

File: include/openmc/simulation.h [hunk 2]
```diff
@@ -32,16 +32,20 @@ extern "C" double k_col_abs; //!< sum over batches of k_collision * k_absorption
 extern "C" double
   k_col_tra; //!< sum over batches of k_collision * k_tracklength
 extern "C" double
-  k_abs_tra;               //!< sum over batches of k_absorption * k_tracklength
+  k_abs_tra; //!< sum over batches of k_absorption * k_tracklength
```
Comment (GiudGiud) [inline include/openmc/simulation.h:35]:
```suggestion
  k_abs_tra;               //!< sum over batches of k_absorption * k_tracklength
```

File: include/openmc/simulation.h [hunk 3]
```diff
@@ -32,16 +32,20 @@ extern "C" double k_col_abs; //!< sum over batches of k_collision * k_absorption
...
+extern bool
+  migration_present;       //! Does an active tally contains a migration score
```
Comment (GiudGiud) [inline include/openmc/simulation.h:37]: this global variable is justifiable because we need it during the simulation at every BC

File: include/openmc/simulation.h [hunk 4]
```diff
@@ -32,16 +32,20 @@ extern "C" double k_col_abs; //!< sum over batches of k_collision * k_absorption
...
 extern "C" int n_lost_particles;   //!< cumulative number of lost particles
 extern "C" bool need_depletion_rx; //!< need to calculate depletion rx?
-extern "C" int restart_batch;      //!< batch at which a restart job resumed
-extern "C" bool satisfy_triggers;  //!< have tally triggers been satisfied?
-extern int ssw_current_file;       //!< current surface source file
-extern "C" int total_gen;          //!< total number of generations simulated
-extern double total_weight;        //!< Total source weight in a batch
-extern int64_t work_per_rank;      //!< number of particles per MPI rank
+extern bool
+  nonvacuum_boundary_present;     //!< Does the geometry contain non-vacuum b.c.
```
Comment (GiudGiud) [inline include/openmc/simulation.h:42]: this one is not, it's a setup item it should be handled with a check over all surfaces not the setting of a global boolean

File: src/tallies/tally.cpp [hunk 5]
```diff
@@ -601,6 +607,19 @@ void Tally::set_scores(const vector<std::string>& scores)
         estimator_ = TallyEstimator::ANALOG;
       break;
 
+    case SCORE_MIGRATION:
+      if (estimator_ != TallyEstimator::TRACKLENGTH)
+        fatal_error(
+          "Migration-area can only be tallies with tracklength estimator");
```
Comment (GiudGiud) [inline src/tallies/tally.cpp]:
```suggestion
          "Migration-area can only be tallied with a tracklength estimator");
```

File: tests/regression_tests/migration/test.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    # Tally 
+    groups = openmc.mgxs.EnergyGroups('CASMO-70')
+    energy_filter = openmc.EnergyFilter(groups.group_edges)
+    tally = openmc.Tally()
+    tally.scores = ["migration-area", "flux", "total"]
+    tally.filters = [energy_filter]
+    model.tallies = [tally]
+
+    return model
```
Comment (GiudGiud) [inline tests/regression_tests/migration/test.py:39]: is this test case the one used for the figures? If so let's keep it, we somewhat know what the result looks like

But we also need coverage on all the other types of supported boundary conditions. So let's make either:
- a dummy test featuring all these
or better:
- a test featuring these BCs but that gives the exact same result due to the symmetries of the geometry

File: src/boundary_condition.cpp [hunk 7]
```diff
@@ -43,6 +44,15 @@ void ReflectiveBC::handle_particle(Particle& p, const Surface& surf) const
   // Handle the effects of the surface albedo on the particle's weight.
   BoundaryCondition::handle_albedo(p, surf);
 
+  // Handle phantom birth location if migration present
+  if (simulation::migration_present) {
+    auto r = p.r();
+    auto n = surf.normal(r);
+    n /= n.norm();
+    auto r_born = p.r_born();
```
Comment (GiudGiud) [inline src/boundary_condition.cpp]:
```suggestion
    const auto & r = p.r();
    auto n = surf.normal(r);
    n /= n.norm();
    auto & r_born = p.r_born();
```

File: src/boundary_condition.cpp [hunk 8]
```diff
@@ -244,6 +263,18 @@ void RotationalPeriodicBC::handle_particle(
   // Handle the effects of the surface albedo on the particle's weight.
   BoundaryCondition::handle_albedo(p, surf);
 
+  // Handle phantom birth location if migration present
+  if (simulation::migration_present) {
+    auto r_born = p.r_born();
```
Comment (GiudGiud) [inline src/boundary_condition.cpp]:
```suggestion
    auto & r_born = p.r_born();
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: include/openmc/simulation.h (hunk 2 above)
Comment (GuySten) [reply include/openmc/simulation.h:35]: This change is not compatible with clang-format.

File: include/openmc/simulation.h (hunk 4 above)
Comment (GuySten) [reply include/openmc/simulation.h:42]: This variable is set when reading surfaces and used when setting tallies.
So removing it will involve alot of function signature change.
I think this way is better.

Comment (GiudGiud) [reply include/openmc/simulation.h:42]: you would just need to not set it when reading surfaces, then when setting tallies you can loop over the geometry/surfaces if migration tallies are used to check their type

Comment (GuySten) [reply include/openmc/simulation.h:42]: In general it is not recommended to iterate over cells or surfaces twice because there are models with alot of cells and surfaces.

Comment (GuySten) [reply include/openmc/simulation.h:42]: @paulromano, what do you think?

This issue can affect performance.

**Comment by GiudGiud** [comment]
@GuySten you should be able to lump the suggestions from the files tab (cant do it from the conversation tab) to limit the test suite runs (for next time)

File: tests/regression_tests/migration/test.py (hunk 6 above)
Comment (GuySten) [reply tests/regression_tests/migration/test.py:39]: The current regression test is the same one from the example.

I've added a test that a large water sphere gives the same results as a small one with reflective boundary conditions.

I think I know how to extend this PR to white boundary conditions but I will leave that as a potential future PR.

Comment (GiudGiud) [reply tests/regression_tests/migration/test.py:39]: > I think I know how to extend this PR to white boundary conditions but I will leave that as a potential future PR.

Yes for sure.
If OpenMC testing involves exception testing we'll just want a test that shows the error.
@paulromano might give guidance on whether it's needed here

File: include/openmc/simulation.h (hunk 4 above)
Comment (GiudGiud) [reply include/openmc/simulation.h:42]: Iterating once, or even multiple times but outside of the particle/event loop, on every surface is cheap. 

You should implement the changes so we can move forward here. Paul likely does not have time to review

Comment (paulromano) [reply include/openmc/simulation.h:42]: I do plan to review this. As always, I do appreciate the contribution and your continued patience as I work through the long backlog of PRs.

**Review by paulromano — CHANGES_REQUESTED** [review]
Putting a placeholder review and will try to give an actual full review soon

File: include/openmc/simulation.h (hunk 4 above)
Comment (GiudGiud) [reply include/openmc/simulation.h:42]: Sounds good, thanks!

**Comment by GiudGiud** [comment]
@jtramm

File: include/openmc/simulation.h (hunk 4 above)
Comment (GuySten) [reply include/openmc/simulation.h]: I discovered that I used the nonvacuum variable wrongly because it did not read dagmc surfaces.
So I removed it in the last commit.

File: tests/regression_tests/migration/test.py (hunk 6 above)
Comment (GuySten) [reply tests/regression_tests/migration/test.py:39]: I've figured out that I actually don't know how to extend this PR to either white bcs or bcs with albedo.
So i now check before running that they are not present. 
I've also introduced additional tests.
